gv.c

```diff
@@ -1696,15 +1697,73 @@ reasons.
 
 HV*
 Perl_gv_stashsvpvn_cached(pTHX_ SV *namesv, const char *name, U32 namelen, I32 flags)
+{
+    PERL_ARGS_ASSERT_GV_STASHSVPVN_CACHED;
+    if(namesv)
+        return Perl_gv_stashsvpvn_cached_p(aTHX_ flags | GVCf_ISSV, (void*)namesv);
+    else
+        return Perl_gv_stashsvpvn_cached_p(aTHX_ flags | GVCf_ISPV | GVCf_HASVA_LEN,
+            (void*)name, namelen);
+}
+
+/* Not public API, "..." takes exactly 1 optional arg, "U32 namelen" for the
+   public facing _pvn() variant. All other varieties, pv() pvs() sv() hek()
+   do not use the optional 3rd arg.*/
+HV*
+Perl_gv_stashsvpvn_cached_p(pTHX_ I32 flags, void * namevp, ...)
 {
     HV* stash;
     HE* he;
+    SV* namesv;
+    const char* name;
+    va_list args;
+    U32 namelen;
+    U32 hash;
 
-    PERL_ARGS_ASSERT_GV_STASHSVPVN_CACHED;
+    PERL_ARGS_ASSERT_GV_STASHSVPVN_CACHED_P;
+    va_start(args, namevp);
+    if(GV_CACHE_ISRC(flags)) {
+        if(GV_CACHE_ISHEK(flags)) {
+            HEK* hek = (HEK*)namevp;
+            I32 hek_len =  HEK_LEN(hek);
+            if(hek_len == HEf_SVKEY) {
+                namesv = *(SV**)HEK_KEY(hek);
+                namevp = (void*)namesv;
+                goto have_sv;
+            }
+            namesv = NULL;
+            name = HEK_KEY(hek);
+            namelen = hek_len;
+            flags = (flags & ~SVf_UTF8) | (HEK_UTF8(hek) ? SVf_UTF8 : 0);
+/* hv_fetchhek() isn't used here. Its a macro and doesn't currently
+   do the optimisation you think it is supposed to do. Using macro
```

Comment:
What optimization is that?

Reply:
https://github.com/Perl/perl5/blob/49e1c32066791d2d86ecb35bd4de885200e36e80/hv.c#L849